core/src/software/aws/toolkits/core/credentials/sso/SsoAccessTokenProvider.kt

```diff
@@ -0,0 +1,135 @@
+import software.amazon.awssdk.services.ssooidc.model.InvalidClientException
+import software.amazon.awssdk.services.ssooidc.model.SlowDownException
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+
+/**
+ * Takes care of creating/refreshing the SSO access token required to fetch SSO-based credentials.
+ */
+class SsoAccessTokenProvider(
+    private val ssoUrl: String,
+    private val ssoRegion: String,
+    private val onPendingToken: SsoLoginCallback,
+    private val cache: SsoCache,
+    private val client: SsoOidcClient,
+    private val clock: Clock = Clock.systemUTC()
+) {
+    suspend fun accessToken(): AccessToken {
+        cache.loadAccessToken(ssoUrl)?.let {
+            return it
+        }
+
+        val token = pollForToken()
+
+        cache.saveAccessToken(ssoUrl, token)
+
+        return token
+    }
+
+    private fun registerClient(): ClientRegistration {
+        cache.loadClientRegistration(ssoRegion)?.let {
+            return it
+        }
+
+        // Based on botocore: https://github.com/boto/botocore/blob/5dc8ee27415dc97cfff75b5bcfa66d410424e665/botocore/utils.py#L1769
+        val registerResponse = client.registerClient {
+            it.clientType(CLIENT_REGISTRATION_TYPE)
+            it.clientName("aws-toolkit-jetbrains-${Instant.now(clock)}")
+        }
+
+        val registeredClient = ClientRegistration(
+            registerResponse.clientId(),
+            registerResponse.clientSecret(),
+            Instant.ofEpochSecond(registerResponse.clientSecretExpiresAt())
+        )
+
+        cache.saveClientRegistration(ssoRegion, registeredClient)
+
+        return registeredClient
+    }
+
+    private fun authorizeClient(clientId: ClientRegistration): Authorization {
+        // Should not be cached, only good for 1 token and short lived
+        val authorizationResponse = try {
+            client.startDeviceAuthorization {
+                it.startUrl(ssoUrl)
+                it.clientId(clientId.clientId)
+                it.clientSecret(clientId.clientSecret)
+            }
+        } catch (e: InvalidClientException) {
+            cache.invalidateClientRegistration(ssoRegion)
+            throw e
+        }
+
+        return Authorization(
+            authorizationResponse.deviceCode(),
+            authorizationResponse.userCode(),
+            authorizationResponse.verificationUri(),
+            authorizationResponse.verificationUriComplete(),
+            Instant.now(clock).plusSeconds(authorizationResponse.expiresIn().toLong()),
+            authorizationResponse.interval()?.toLong()
+                ?: DEFAULT_INTERVAL
+        )
+    }
+
+    private suspend fun pollForToken(): AccessToken {
+        val registration = registerClient()
+        val authorization = authorizeClient(registration)
+
+        onPendingToken.tokenPending(authorization)
+
+        var backOffTime = Duration.ofSeconds(authorization.pollInterval)
+
+        while (true) {
```

Comment:
do we need a max retry count?

Reply:
Nope, we aren't supposed to have one. We are to respect the expiresIn / go until the service says do not keep trying. They throw `ExpiredTokenException` and we land in the catch all block